lnwallet/fee_estimator.go

```diff
@@ -524,3 +542,205 @@ func (e ExternalFeeSource) ParseResponse(r io.Reader) (map[uint32]uint32, error)
 var _ WebAPIFeeSource = (*ExternalFeeSource)(nil)
+
+// WebAPIFeeEstimator is an implementation of the FeeEstimator interface that
+// queries an HTTP-based fee estimation from an existing web API.
+type WebAPIFeeEstimator struct {
+	started uint32 // To be used atomically.
+	stopped uint32 // To be used atomically.
+
+	// apiSource is the backing web API source we'll use for our queries.
+	apiSource WebAPIFeeSource
+
+	// chainNotifier is used by the WebAPIFeeEstimator to determine when to
+	// re-query the API for fresh fees. We re-query whenever a new block
+	// comes in.
+	chainNotifier chainntnfs.ChainNotifier
+
+	// blockEpochs is our block notification subscription.
+	blockEpochs *chainntnfs.BlockEpochEvent
+
+	// feeByBlockTarget is our cache for fees pulled from the API. When a
+	// fee estimate request comes in, we pull the estimate from this array
+	// rather than re-querying the API, to prevent an inadvertent DoS attack.
+	feesMtx          sync.Mutex
+	feeByBlockTarget map[uint32]uint32
+
+	// defaultFeePerKw is a fallback value that we'll use if we're unable
+	// to query the API for any reason.
+	defaultFeePerKw SatPerKWeight
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// NewWebAPIFeeEstimator creates a new WebAPIFeeEstimator from a given URL and a
+// fallback default fee. The fees are updated whenever a new block is mined.
+func NewWebAPIFeeEstimator(
+	api WebAPIFeeSource, notifier chainntnfs.ChainNotifier, defaultFee SatPerKWeight) *WebAPIFeeEstimator {
+
+	return &WebAPIFeeEstimator{
+		apiSource:        api,
+		chainNotifier:    notifier,
+		feeByBlockTarget: make(map[uint32]uint32),
+		defaultFeePerKw:  defaultFee,
+		quit:             make(chan struct{}),
+	}
+}
+
+// EstimateFeePerKW takes in a target for the number of blocks until an initial
+// confirmation and returns the estimated fee expressed in sat/kw.
+//
+// NOTE: This method is part of the FeeEstimator interface.
+func (w *WebAPIFeeEstimator) EstimateFeePerKW(numBlocks uint32) (SatPerKWeight, error) {
+	if numBlocks > MaxBlockTarget {
+		numBlocks = MaxBlockTarget
+	} else if numBlocks < MinBlockTarget {
+		return 0, fmt.Errorf("conf target of %v is too low, minimum "+
+			"accepted is %v", numBlocks, MinBlockTarget)
+	}
+
+	w.feesMtx.Lock()
+	feePerKb, err := w.getCachedFee(numBlocks)
+	w.feesMtx.Unlock()
+	if err != nil {
+		return 0, err
+	}
+
+	// If the result is too low, then we'll clamp it to our current fee
+	// floor.
+	satPerKw := SatPerKVByte(feePerKb).FeePerKWeight()
+	if satPerKw < FeePerKwFloor {
+		satPerKw = FeePerKwFloor
+	}
+
+	walletLog.Debugf("Web API returning %v sat/kw for conf target of %v",
+		int64(satPerKw), numBlocks)
+
+	return satPerKw, nil
+}
+
+// Start signals the FeeEstimator to start any processes or goroutines it needs
+// to perform its duty.
+//
+// NOTE: This method is part of the FeeEstimator interface.
+func (w *WebAPIFeeEstimator) Start() error {
+	if !atomic.CompareAndSwapUint32(&w.started, 0, 1) {
```

Comment:
We've largely moved to using a [`sync.Once`](https://golang.org/pkg/sync/#Once) across the codebase.

Reply:
fixed